src/parser/literal.rs

```diff
@@ -67,6 +72,115 @@ pub fn parse_guid(inp: &str) -> IResult<&str, String> {
     Ok((i, guid_str.to_string()))
 }
 
+pub fn parse_year(inp: &str) -> IResult<&str, i32> {
+    let parser = recognize(tuple((opt(char('-')), take_while_m_n(4, 4, is_digit))));
+
+    // Infallible, as 4 digits always fit an i32
+    map(parser, |s: &str| s.parse::<i32>().unwrap())(inp)
+}
+
+pub fn n_digits_between(inp: &str, n_digits: usize, min: u8, max: u8) -> IResult<&str, u8> {
+    let digits = take_while_m_n(n_digits, n_digits, is_digit);
+
+    verify(
+        map(digits, |s: &str| s.parse::<u8>().unwrap()),
+        |val: &u8| val >= &min && val <= &max,
+    )(inp)
+}
+
+pub fn parse_month(inp: &str) -> IResult<&str, Month> {
+    map_res(
+        |i| n_digits_between(i, 2, 1, 12),
+        |val| Month::try_from(val),
+    )(inp)
+}
+
+pub fn parse_day(inp: &str) -> IResult<&str, u8> {
+    n_digits_between(inp, 2, 1, 31)
+}
+
+pub fn parse_date(inp: &str) -> IResult<&str, Date> {
+    // OData `year`s can be negative, conflicting with ISO8601.
+    // So we don't use `time::*::parse`
+    let parser = tuple((parse_year, char('-'), parse_month, char('-'), parse_day));
+
+    map_res(parser, |(y, _, m, _, d)| Date::from_calendar_date(y, m, d))(inp)
+}
+
+pub fn parse_hour(inp: &str) -> IResult<&str, u8> {
+    n_digits_between(inp, 2, 0, 24)
+}
+
+pub fn parse_minute(inp: &str) -> IResult<&str, u8> {
+    n_digits_between(inp, 2, 0, 59)
+}
+
+pub fn parse_fractional_seconds(inp: &str) -> IResult<&str, u32> {
+    // Parses the "fractionalSeconds" after the dot to an amount expressed in
+    // nanoseconds
+    let digits = take_while_m_n(1, 12, is_digit);
+
+    // Since we want to express as nanoseconds, and this is the fractional part
+    // of a number, we ensure we have exactly 9 digits before parsing:
+    let nanos = map(digits, |d: &str| format!("{d:0<9.9}"));
```

Comment:
This feels like such a stupid hack, but my fried-day brain can't see the "correct" way.